File: Client/src/dtos/configurations/configurationDto.ts
export enum ConfigurationDiscriminator {
  TemperatureConfiguration = "temperature",
  ColorConfiguration = "color",
}

export interface ConfigurationDto {
  $type: ConfigurationDiscriminator;
  id: string;
  name: string;
  devicePath: string;
  applyBrightness: boolean;
  brightness: number;
}

File: Server/DTOs/Configurations/ConfigurationDto.cs
using Swashbuckle.AspNetCore.Annotations;
using System.Text.Json.Serialization;

namespace ScreenTemperature.DTOs.Configurations
{
    [SwaggerDiscriminator("$type")]
    [SwaggerSubType(typeof(TemperatureConfigurationDto), DiscriminatorValue = "temperature")]
    [JsonDerivedType(typeof(TemperatureConfigurationDto), "temperature")]
    [SwaggerSubType(typeof(ColorConfigurationDto), DiscriminatorValue = "color")]
    [JsonDerivedType(typeof(ColorConfigurationDto), "color")]
    public abstract class ConfigurationDto
    {
        public Guid Id { get; set; }
        public string Name { get; set; }
        public string DevicePath { get; set; }
        public bool ApplyBrightness { get; set; }
        public byte Brightness { get; set; }
        public bool ApplyAtStartup { get; set; }
    }
}
